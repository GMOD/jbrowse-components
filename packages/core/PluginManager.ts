import {
  types,
  IAnyType,
  IAnyModelType,
  isModelType,
  isType,
} from 'mobx-state-tree'

import PluggableElementBase from './pluggableElementTypes/PluggableElementBase'
import RendererType from './pluggableElementTypes/renderers/RendererType'
import AdapterType from './pluggableElementTypes/AdapterType'
import TrackType from './pluggableElementTypes/TrackType'
import DisplayType from './pluggableElementTypes/DisplayType'
import ViewType from './pluggableElementTypes/ViewType'
import WidgetType from './pluggableElementTypes/WidgetType'
import ConnectionType from './pluggableElementTypes/ConnectionType'
import RpcMethodType from './pluggableElementTypes/RpcMethodType'

import {
  ConfigurationSchema,
  isBareConfigurationSchemaType,
} from './configuration'

import Plugin from './Plugin'
import ReExports from './ReExports'

import {
  PluggableElementType,
  PluggableElementMember,
} from './pluggableElementTypes'
import { AnyConfigurationSchemaType } from './configuration/configurationSchema'
import { AbstractRootModel } from './util'
import CorePlugin from './CorePlugin'

/** little helper class that keeps groups of callbacks that are
then run in a specified order by group */
class PhasedScheduler<PhaseName extends string> {
  phaseCallbacks = new Map<PhaseName, Function[]>()

  phaseOrder: PhaseName[] = []

  constructor(...phaseOrder: PhaseName[]) {
    this.phaseOrder = phaseOrder
  }

  add(phase: PhaseName, callback: Function) {
    if (!this.phaseOrder.includes(phase)) {
      throw new Error(`unknown phase ${phase}`)
    }
    let phaseCallbacks = this.phaseCallbacks.get(phase)
    if (!phaseCallbacks) {
      phaseCallbacks = []
      this.phaseCallbacks.set(phase, phaseCallbacks)
    }
    phaseCallbacks.push(callback)
  }

  run() {
    this.phaseOrder.forEach(phaseName => {
      const phaseCallbacks = this.phaseCallbacks.get(phaseName)
      if (phaseCallbacks) {
        phaseCallbacks.forEach(callback => callback())
      }
    })
  }
}

type PluggableElementTypeGroup =
  | 'renderer'
  | 'adapter'
  | 'display'
  | 'track'
  | 'connection'
  | 'view'
  | 'widget'
  | 'rpc method'

/** internal class that holds the info for a certain element type */
class TypeRecord<ElementClass extends PluggableElementBase> {
  registeredTypes: { [name: string]: ElementClass } = {}

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  baseClass: { new (...args: any[]): ElementClass }

  typeName: string

  constructor(
    typeName: string,
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    elementType: { new (...args: any[]): ElementClass },
  ) {
    this.typeName = typeName
    this.baseClass = elementType
  }

  add(name: string, t: ElementClass) {
    this.registeredTypes[name] = t
  }

  has(name: string) {
    return name in this.registeredTypes
  }

  get(name: string) {
    if (!this.has(name))
      throw new Error(
        `${this.typeName} '${name}' not found, perhaps its plugin is not loaded or its plugin has not added it.`,
      )
    return this.registeredTypes[name]
  }

  all() {
    return Object.values(this.registeredTypes)
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFunction = (...args: any) => any

export default class PluginManager {
  plugins: Plugin[] = []

  elementCreationSchedule = new PhasedScheduler<PluggableElementTypeGroup>(
    'renderer',
    'adapter',
    'display',
    'track',
    'connection',
    'view',
    'widget',
    'rpc method',
  )

  rendererTypes = new TypeRecord('RendererType', RendererType)

  adapterTypes = new TypeRecord('AdapterType', AdapterType)

  trackTypes = new TypeRecord('TrackType', TrackType)

  displayTypes = new TypeRecord('DisplayType', DisplayType)

  connectionTypes = new TypeRecord('ConnectionType', ConnectionType)

  viewTypes = new TypeRecord('ViewType', ViewType)

  widgetTypes = new TypeRecord('WidgetType', WidgetType)

  rpcMethods = new TypeRecord('RpcMethodType', RpcMethodType)

  configured = false

  rootModel?: AbstractRootModel

  constructor(initialPlugins: Plugin[] = []) {
    // add the core plugin
    this.addPlugin(new CorePlugin())

    // add all the initial plugins
    initialPlugins.forEach(plugin => {
      this.addPlugin(plugin)
    })
  }

  addPlugin(plugin: Plugin) {
    if (this.configured) {
      throw new Error('JBrowse already configured, cannot add plugins')
    }
    if (this.plugins.includes(plugin)) {
      throw new Error('plugin already installed')
    }
    plugin.install(this)
    this.plugins.push(plugin)
    return this
  }

  getPlugin(name: string) {
    return this.plugins.find(p => p.name === name)
  }

  hasPlugin(name: string) {
    return this.getPlugin(name) !== undefined
  }

  createPluggableElements() {
    // run the creation callbacks for each element type in order.
    // see elementCreationSchedule above for the creation order
    this.elementCreationSchedule.run()
    delete this.elementCreationSchedule
    return this
  }

  setRootModel(rootModel: AbstractRootModel) {
    this.rootModel = rootModel
  }

  configure() {
    if (this.configured) throw new Error('already configured')

    this.plugins.forEach(plugin => plugin.configure(this))

    this.configured = true

    // console.log(JSON.stringify(getSnapshot(model)))

    return this
  }

  getElementTypeRecord(
    groupName: PluggableElementTypeGroup,
  ): TypeRecord<PluggableElementBase> {
    switch (groupName) {
      case 'adapter':
        return this.adapterTypes
      case 'connection':
        return this.connectionTypes
      case 'widget':
        return this.widgetTypes
      case 'renderer':
        return this.rendererTypes
      case 'display':
        return this.displayTypes
      case 'track':
        return this.trackTypes
      case 'view':
        return this.viewTypes
      case 'rpc method':
        return this.rpcMethods
      default:
        throw new Error(`invalid element type '${groupName}'`)
    }
  }

  addElementType(
    groupName: PluggableElementTypeGroup,
    creationCallback: (pluginManager: PluginManager) => PluggableElementType,
  ) {
    if (typeof creationCallback !== 'function') {
      throw new Error(
        'must provide a callback function that returns the new type object',
      )
    }
    const typeRecord = this.getElementTypeRecord(groupName)

    this.elementCreationSchedule.add(groupName, () => {
      const newElement = creationCallback(this)
      if (!newElement.name)
        throw new Error(`cannot add a ${groupName} with no name`)

      if (typeRecord.has(newElement.name)) {
        throw new Error(
          `${groupName} ${newElement.name} already registered, cannot register it again`,
        )
      }

      typeRecord.add(newElement.name, newElement)
    })

    return this
  }

  getElementType(groupName: PluggableElementTypeGroup, typeName: string) {
    const typeRecord = this.getElementTypeRecord(groupName)
    return typeRecord.get(typeName)
  }

  getElementTypesInGroup(groupName: PluggableElementTypeGroup) {
    const typeRecord = this.getElementTypeRecord(groupName)
    return typeRecord.all()
  }

  /** get a MST type for the union of all specified pluggable MST types */
  pluggableMstType(
    typeGroup: PluggableElementTypeGroup,
    fieldName: PluggableElementMember,
    fallback: IAnyType = types.maybe(types.null),
  ) {
    const pluggableTypes: IAnyModelType[] = []
    this.getElementTypeRecord(typeGroup)
      .all()
      .forEach(t => {
        const thing = t[fieldName]
        if (isType(thing) && isModelType(thing)) {
          pluggableTypes.push(thing)
        }
      })
    // try to smooth over the case when no types are registered, mostly encountered in tests
    if (pluggableTypes.length === 0) {
      console.warn(
        `No JBrowse pluggable types found matching ('${typeGroup}','${fieldName}')`,
      )
      return fallback
    }
    return types.union(...pluggableTypes)
  }

  /** get a MST type for the union of all specified pluggable config schemas */
  pluggableConfigSchemaType(
    typeGroup: PluggableElementTypeGroup,
    fieldName: PluggableElementMember = 'configSchema',
  ) {
    const pluggableTypes: AnyConfigurationSchemaType[] = []
    this.getElementTypeRecord(typeGroup)
      .all()
      .forEach(t => {
        const thing = t[fieldName]
        if (isBareConfigurationSchemaType(thing)) {
          pluggableTypes.push(thing)
        }
      })
    if (pluggableTypes.length === 0)
      pluggableTypes.push(ConfigurationSchema('Null', {}))
    return types.union(...pluggableTypes)
  }

  jbrequireCache = new Map()

  lib = ReExports

  load = <FTYPE extends AnyFunction>(lib: FTYPE): ReturnType<FTYPE> => {
    if (!this.jbrequireCache.has(lib)) this.jbrequireCache.set(lib, lib(this))
    return this.jbrequireCache.get(lib)
  }

  /**
   * Get the re-exported version of the given package name.
   * Throws an error if the package is not re-exported by the plugin manager.
   *
   * @returns the library's default export
   */
  jbrequire = (
    lib: keyof typeof ReExports | AnyFunction | { default: AnyFunction },
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ): any => {
    if (typeof lib === 'string') {
      const pack = this.lib[lib]
      if (!pack) {
        throw new TypeError(
          `No jbrequire re-export defined for package '${lib}'. If this package must be shared between plugins, add it to ReExports.js. If it does not need to be shared, just import it normally.`,
        )
      }
      return pack
    }

    if (typeof lib === 'function') {
      return this.load(lib)
    }

    if (lib.default) return this.jbrequire(lib.default)

    throw new TypeError(
      'lib passed to jbrequire must be either a string or a function',
    )
  }

  getRendererType(typeName: string): RendererType {
    return this.rendererTypes.get(typeName)
  }

  getRendererTypes(): RendererType[] {
    return this.rendererTypes.all()
  }

  getAdapterType(typeName: string): AdapterType {
    return this.adapterTypes.get(typeName)
  }

  getTrackType(typeName: string): TrackType {
    return this.trackTypes.get(typeName)
  }

  getDisplayType(typeName: string): DisplayType {
    return this.displayTypes.get(typeName)
  }

  getViewType(typeName: string): ViewType {
    return this.viewTypes.get(typeName)
  }

  getWidgetType(typeName: string): WidgetType {
    return this.widgetTypes.get(typeName)
  }

  getConnectionType(typeName: string): ConnectionType {
    return this.connectionTypes.get(typeName)
  }

  getRpcMethodType(methodName: string): RpcMethodType {
    return this.rpcMethods.get(methodName)
  }

  addRendererType(
    creationCallback: (pluginManager: PluginManager) => RendererType,
  ): this {
    return this.addElementType('renderer', creationCallback)
  }

  addAdapterType(
    creationCallback: (pluginManager: PluginManager) => AdapterType,
  ): this {
    return this.addElementType('adapter', creationCallback)
  }

  addTrackType(
    creationCallback: (pluginManager: PluginManager) => TrackType,
  ): this {
    // Goes through the already-created displays and registers the ones that
    // specify this track type
    const callback = () => {
      const track = creationCallback(this)
      ;(this.getElementTypesInGroup('display') as DisplayType[]).forEach(
        display => {
          if (
            display.trackType === track.name &&
            // track may have already added the displayType in its creationCallback
            !track.displayTypes.includes(display)
          ) {
            track.addDisplayType(display)
          }
        },
      )
      return track
    }
    return this.addElementType('track', callback)
  }

  addDisplayType(
    creationCallback: (pluginManager: PluginManager) => DisplayType,
  ): this {
    return this.addElementType('display', creationCallback)
  }

  addViewType(
    creationCallback: (pluginManager: PluginManager) => ViewType,
  ): this {
    const callback = () => {
      const newView = creationCallback(this)
      ;(this.getElementTypesInGroup('display') as DisplayType[]).forEach(
        display => {
          if (
            display.viewType === newView.name &&
            // view may have already added the displayType in its creationCallback
            !newView.displayTypes.includes(display)
          ) {
            newView.addDisplayType(display)
          }
        },
      )
      return newView
    }
    return this.addElementType('view', callback)
  }

  addWidgetType(
    creationCallback: (pluginManager: PluginManager) => WidgetType,
  ): this {
    return this.addElementType('widget', creationCallback)
  }

  addConnectionType(
    creationCallback: (pluginManager: PluginManager) => ConnectionType,
  ): this {
    return this.addElementType('connection', creationCallback)
  }

  addRpcMethod(
    creationCallback: (pluginManager: PluginManager) => RpcMethodType,
  ): this {
    return this.addElementType('rpc method', creationCallback)
  }
}
