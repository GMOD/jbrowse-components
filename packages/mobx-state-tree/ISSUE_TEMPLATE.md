**If your issue isn't a bug report, please consider using discussion threads instead of opening an issue: https://github.com/mobxjs/mobx-state-tree/discussions**
