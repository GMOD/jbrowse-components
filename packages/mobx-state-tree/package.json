{
  "name": "@jbrowse/mobx-state-tree",
  "version": "5.4.2",
  "description": "Opinionated, transactional, MobX powered state container",
  "main": "dist/index.js",
  "module": "esm/index.js",
  "sideEffects": false,
  "scripts": {
    "clean": "rimraf dist esm",
    "prebuild": "yarn clean",
    "build:esm": "tsup  --format esm src/index.ts --dts -d esm",
    "build:es5": "tsup  src/index.ts --dts",
    "build": "yarn build:es5 && yarn build:esm",
    "typecheck": "tsc --noEmit"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/mobxjs/@jbrowse/@jbrowse/mobx-state-tree.git"
  },
  "author": "Michel Weststrate",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/mobxjs/mobx-state-tree/issues"
  },
  "files": [
    "dist/"
  ],
  "peerDependencies": {
    "mobx": "^6.3.0"
  },
  "keywords": [
    "mobx",
    "mobx-state-tree",
    "promise",
    "reactive",
    "frp",
    "functional-reactive-programming",
    "state management"
  ],
  "dependencies": {
    "ts-essentials": "^10.0.4"
  },
  "devDependencies": {
    "@swc/core": "^1.11.20",
    "tsup": "^8.4.0"
  }
}
