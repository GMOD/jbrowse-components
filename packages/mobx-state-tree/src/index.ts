// tslint:disable-next-line:no_unused-variable
import { IObservableArray, ObservableMap } from 'mobx'

/* all code is initially loaded through internal, to avoid circular dep issues */
export {
  typecheck,
  escapeJsonPath,
  unescapeJsonPath,
  joinJsonPath,
  splitJsonPath,
  decorate,
  addMiddleware,
  process,
  isStateTreeNode,
  flow,
  castFlowReturn,
  applyAction,
  onAction,
  recordActions,
  createActionTrackingMiddleware,
  createActionTrackingMiddleware2,
  setLivelinessChecking,
  getLivelinessChecking,
  setLivelynessChecking, // to be deprecated
  getType,
  getChildType,
  onPatch,
  onSnapshot,
  applyPatch,
  recordPatches,
  protect,
  unprotect,
  isProtected,
  applySnapshot,
  getSnapshot,
  hasParent,
  getParent,
  hasParentOfType,
  getParentOfType,
  getRoot,
  getPath,
  getPathParts,
  isRoot,
  resolvePath,
  resolveIdentifier,
  getIdentifier,
  tryResolve,
  getRelativePath,
  clone,
  detach,
  destroy,
  isAlive,
  addDisposer,
  getEnv,
  walk,
  getMembers,
  getPropertyMembers,
  cast,
  castToSnapshot,
  castToReferenceSnapshot,
  isType,
  isArrayType,
  isFrozenType,
  isIdentifierType,
  isLateType,
  isLiteralType,
  isMapType,
  isModelType,
  isOptionalType,
  isPrimitiveType,
  isReferenceType,
  isRefinementType,
  isUnionType,
  tryReference,
  isValidReference,
  getNodeId,
  getRunningActionContext,
  isActionContextChildOf,
  isActionContextThisOrChildOf,
  types,
  types as t, // We do this as a less ambiguous term for the traditional types module, which sometimes gets confused when discussing "types" in general
  toGeneratorFunction,
  toGenerator,
} from './internal'

export type {
  IModelType,
  IAnyModelType,
  IDisposer,
  IMSTMap,
  IMapType,
  IMSTArray,
  IArrayType,
  IType,
  IAnyType,
  ModelPrimitive,
  ISimpleType,
  IComplexType,
  IAnyComplexType,
  IReferenceType,
  IStateTreeNode,
  IAnyStateTreeNode,
  IActionRecorder,
  ISerializedActionCall,
  ReferenceIdentifier,
  ISnapshotProcessor,
  ISnapshotProcessors,
  OnReferenceInvalidated,
  OnReferenceInvalidatedEvent,
  ReferenceOptions,
  ReferenceOptionsGetSet,
  ReferenceOptionsOnInvalidated,
  IActionContext,
  TypeOfValue,
  IModelReflectionData,
  IModelReflectionPropertiesData,
  IMaybeIType,
  IMaybe,
  IMaybeNull,
  IOptionalIType,
  OptionalDefaultValueOrFunction,
  ValidOptionalValue,
  ValidOptionalValues,
  LivelynessMode, // to be deprecated
  ModelSnapshotType,
  ModelCreationType,
  ModelSnapshotType2,
  ModelCreationType2,
  ModelInstanceType,
  ModelInstanceTypeProps,
  ModelPropertiesDeclarationToProperties,
  ModelProperties,
  ModelPropertiesDeclaration,
  ModelActions,
  ITypeUnion,
  CustomTypeOptions,
  UnionOptions,
  Instance,
  SnapshotIn,
  SnapshotOut,
  SnapshotOrInstance,
  TypeOrStateTreeNodeToStateTreeNode,
  UnionStringArray,
  IMiddlewareEvent,
  IActionTrackingMiddleware2Call,
  IMiddlewareHandler,
  IMiddlewareEventType,
  IActionTrackingMiddlewareHooks,
  IActionTrackingMiddleware2Hooks,
  IJsonPatch,
  IReversibleJsonPatch,
  _CustomCSProcessor,
  _CustomOrOther,
  _CustomJoin,
  _NotCustomized,
  LivelinessMode,
  IPatchRecorder,
} from './internal'
