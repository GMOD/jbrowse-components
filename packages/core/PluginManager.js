import React from 'react'
import * as mst from 'mobx-state-tree'

import RendererType from './pluggableElementTypes/renderers/RendererType'
import AdapterType from './pluggableElementTypes/AdapterType'
import TrackType from './pluggableElementTypes/TrackType'
import ViewType from './pluggableElementTypes/ViewType'
import DrawerWidgetType from './pluggableElementTypes/DrawerWidgetType'
import MenuBarType from './pluggableElementTypes/MenuBarType'
import ConnectionType from './pluggableElementTypes/ConnectionType'

import { ConfigurationSchema } from './configuration'

// little helper class that keeps groups of callbacks that are
// then run in a specified order by group
class PhasedScheduler {
  phaseCallbacks = {}

  constructor(...phaseOrder) {
    this.phaseOrder = phaseOrder
  }

  add(phase, callback) {
    if (!this.phaseOrder.includes(phase)) {
      throw new Error(`unknown phase ${phase}`)
    }
    if (!this.phaseCallbacks[phase]) this.phaseCallbacks[phase] = []
    this.phaseCallbacks[phase].push(callback)
  }

  run() {
    this.phaseOrder.forEach(phaseName => {
      if (this.phaseCallbacks[phaseName]) {
        this.phaseCallbacks[phaseName].forEach(callback => callback())
      }
    })
  }
}

export default class PluginManager {
  plugins = []

  elementTypes = {}

  elementCreationSchedule = new PhasedScheduler(
    'renderer',
    'adapter',
    'track',
    'connection',
    'view',
    'drawer widget',
    'menu bar',
  )

  typeBaseClasses = {
    renderer: RendererType,
    adapter: AdapterType,
    track: TrackType,
    connection: ConnectionType,
    view: ViewType,
    'drawer widget': DrawerWidgetType,
    'menu bar': MenuBarType,
  }

  static lib = { 'mobx-state-tree': mst, React }

  constructor(initialPlugins = []) {
    this.lib = PluginManager.lib

    this.getRendererType = this.getElementType.bind(this, 'renderer')
    this.getAdapterType = this.getElementType.bind(this, 'adapter')
    this.getTrackType = this.getElementType.bind(this, 'track')
    this.getViewType = this.getElementType.bind(this, 'view')
    this.getDrawerWidgetType = this.getElementType.bind(this, 'drawer widget')
    this.getMenuBarType = this.getElementType.bind(this, 'menu bar')
    this.getConnectionType = this.getElementType.bind(this, 'connection')
    this.addRendererType = this.addElementType.bind(this, 'renderer')
    this.addAdapterType = this.addElementType.bind(this, 'adapter')
    this.addTrackType = this.addElementType.bind(this, 'track')
    this.addViewType = this.addElementType.bind(this, 'view')
    this.addDrawerWidgetType = this.addElementType.bind(this, 'drawer widget')
    this.addMenuBarType = this.addElementType.bind(this, 'menu bar')
    this.addConnectionType = this.addElementType.bind(this, 'connection')

    // add all the initial plugins
    initialPlugins.forEach(plugin => {
      this.addPlugin(plugin)
    })
  }

  addPlugin(plugin) {
    if (this.configured) {
      throw new Error('JBrowse already configured, cannot add plugins')
    }
    if (this.plugins.includes(plugin)) {
      throw new Error('plugin already installed')
    }
    // if (!plugin.install) console.error(plugin)
    plugin.install(this)
    this.plugins.push(plugin)
    return this
  }

  /** get a MST type for the union of all specified pluggable MST types */
  pluggableMstType(
    typeGroup,
    fieldName,
    fallback = mst.types.maybe(mst.types.null),
  ) {
    const pluggableTypes = this.getElementTypeMembers(typeGroup, fieldName)
    // try to smooth over the case when no types are registered, mostly encountered in tests
    if (pluggableTypes.length === 0) return fallback
    return mst.types.union(...pluggableTypes)
  }

  /** get a MST type for the union of all specified pluggable config schemas */
  pluggableConfigSchemaType(
    typeGroup,
    fieldName = 'configSchema',
    fallback = ConfigurationSchema('Null', {}),
  ) {
    return this.pluggableMstType(typeGroup, fieldName, fallback)
  }

  addElementType(groupName, creationCallback) {
    if (typeof creationCallback !== 'function') {
      throw new Error('must provide a callback function')
    }
    const typeBaseClass = this.typeBaseClasses[groupName]
    if (!typeBaseClass) {
      throw new Error(`unknown pluggable element type ${groupName}, cannot add`)
    }
    if (!this.elementTypes[groupName]) this.elementTypes[groupName] = {}

    this.elementCreationSchedule.add(groupName, () => {
      const element = creationCallback(this)
      if (
        groupName === 'adapter' &&
        !element.AdapterClass.capabilities.length
      ) {
        throw new Error(
          `Adapter ${
            element.AdapterClass.name
          } must provide a static property "capabilities" that has at least one entry. See BaseAdapter for an example.`,
        )
      }
      if (this.elementTypes[groupName][element.name]) {
        throw new Error(
          `${groupName} ${
            element.name
          } already registered, cannot register it again`,
        )
      }

      this.elementTypes[groupName][element.name] = element
    })

    return this
  }

  getElementType(groupName, typeName) {
    if (!typeName) {
      throw new Error(`invalid type name "${typeName}"`)
    }
    return (this.elementTypes[groupName] || {})[typeName]
  }

  getElementTypesInGroup(groupName) {
    return Object.values(this.elementTypes[groupName] || {})
  }

  getElementTypeMembers(groupName, memberName) {
    return this.getElementTypesInGroup(groupName)
      .map(t => t[memberName])
      .filter(m => !!m)
  }

  configure() {
    if (this.configured) throw new Error('already configured')

    // run the creation callbacks for each element type in order.
    // see elementCreationSchedule above for the creation order
    this.elementCreationSchedule.run()
    delete this.elementCreationSchedule

    this.plugins.forEach(plugin => plugin.configure())

    this.configured = true

    // console.log(JSON.stringify(getSnapshot(model)))

    return this
  }
}
