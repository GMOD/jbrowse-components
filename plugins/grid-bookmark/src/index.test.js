import PluginManager from '@jbrowse/core/PluginManager'
import Config from '@jbrowse/plugin-config'
import { getSnapshot } from 'mobx-state-tree'
import ThisPlugin from '.'

describe('Grid bookmark', () => {
  let pluginManager

  beforeAll(() => {
    const originalConsoleWarn = console.warn
    console.warn = jest.fn()
    pluginManager = new PluginManager([new ThisPlugin(), new Config()])
    pluginManager.createPluggableElements()
    pluginManager.configure()
    console.warn = originalConsoleWarn
  })

  it("won't add if already added", () => {
    expect(() => pluginManager.addPlugin(new ThisPlugin())).toThrow(
      /JBrowse already configured, cannot add plugins/,
    )
  })

  it('adds grid bookmark widget', () => {
    const GridBookmarkWidget = pluginManager.getWidgetType('GridBookmarkWidget')
    const config = GridBookmarkWidget.configSchema.create({
      type: 'GridBookmarkWidget',
    })
    expect(getSnapshot(config)).toMatchSnapshot()
  })
})
