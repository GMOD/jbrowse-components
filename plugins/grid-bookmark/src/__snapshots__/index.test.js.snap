// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid bookmark adds grid bookmark widget 1`] = `Object {}`;
